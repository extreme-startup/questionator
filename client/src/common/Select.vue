<template>
  <select
    v-bind:class="['selectControl', { selectControlError: error }]"
    v-model="selected"
    :options="options"
    @change="onChange($event.target.value)"
    required
  >
    <option value="" disabled selected hidden>{{ placeholder }}</option>
    <option v-for="option in options" :value="option.id" v-bind:key="option.id">
      {{ option.value }}
    </option>
  </select>
</template>

<style>
.selectControl {
  width: 100%;
  height: 40px;
  margin-bottom: 5px;
  border: none;
  font-size: 16px;
  color: var(--text-color);
  outline: none;
}

.selectControlError {
  border: 1px solid red;
}
</style>

<script>
export default {
  name: 'qSelect',
  props: ['value', 'error', 'options', 'placeholder'],
  data: function() {
    return {
      selected: this.value,
    };
  },
  methods: {
    onChange(value) {
      if (value === '') {
        value = null;
      }
      this.$emit('input', value);
    },
  },
};
</script>
