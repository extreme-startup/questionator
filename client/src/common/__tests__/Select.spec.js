import { shallowMount } from '@vue/test-utils';
import Select from '../Select.vue';

const options = [
  {
    id: 0,
    value: 'option 1',
  },
  {
    id: 1,
    value: 'option 2',
  },
];

describe('Select', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(Select, {
      propsData: {
        options,
        placeholder: 'Select placeholder',
        value: '',
        error: true,
      },
    });
  });

  test('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  test('method onChange emit event with proper param if no option selected', () => {
    wrapper.vm.onChange('');
    expect(wrapper.emitted().input[0]).toEqual([null]);
  });

  test('method onChange emit event with proper param if any option selected', () => {
    wrapper.vm.onChange('option 1');
    expect(wrapper.emitted().input[0]).toEqual(['option 1']);
  });
});
