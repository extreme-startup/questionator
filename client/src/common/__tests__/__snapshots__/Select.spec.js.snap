// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select renders correctly 1`] = `
<select
  class="selectControl selectControlError"
  options="[object Object],[object Object]"
  required="required"
>
  <option
    disabled="disabled"
    hidden="hidden"
    selected="selected"
    value=""
  >
    Select placeholder
  </option>
   
  <option
    value="0"
  >
    
    option 1
  
  </option>
  <option
    value="1"
  >
    
    option 2
  
  </option>
</select>
`;
